File: chapter3/3-2-3.py
N = int(input())
# 数列は空白区切りで与えられる
A = list(map(int, input().split()))

v_xor, v_sum = 0, 0
r = 0
ans = 0
for l in range(N):
    while r < N and v_xor ^ A[r] == v_sum + A[r]:
        v_xor ^= A[r]
        v_sum += A[r]
        r += 1
    ans += r - l

    # 長さが 1 ならば必ず条件を満たすので、長さが正かどうかの場合分けは不要
    v_xor ^= A[l]
    v_sum -= A[l]

print(ans)

File: cpp/chapter3/3-2-3.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>
using namespace std;

int main() {
    int N;
    cin >> N;
    vector<int> A(N);
    for(int i = 0; i < N; i++) {
        cin >> A[i];
    }

    int r = 0;
    long long v_xor = 0, v_sum = 0;
    long long ans = 0;
    for(int l = 0; l < N; l++) {
        while(r < N && (v_xor ^ A[r]) == (v_sum + A[r])) {
            v_xor ^= A[r];
            v_sum += A[r];
            r++;
        }
        ans += r - l;

        // 長さが 1 ならば必ず条件を満たすので、長さが正かどうかの場合分けは不要
        v_xor ^= A[l];
        v_sum -= A[l];
    }
    cout << ans << endl;
    return 0;
}
